lightning/src/ln/peer_handler.rs

```diff
@@ -827,207 +842,218 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
+						($peer: expr, $thing: expr) => {
+							match $thing {
+								Ok(x) => x,
+								Err(e) => {
+									match e.action {
+										msgs::ErrorAction::DisconnectPeer { msg: _ } => {
+											//TODO: Try to push msg
+											log_debug!(self.logger, "Error handling message{}; disconnecting peer with: {}", OptionalFromDebugger(&peer_node_id), e.err);
+											return Err(PeerHandleError{ no_connection_possible: false });
+										},
+										msgs::ErrorAction::IgnoreAndLog(level) => {
+											log_given_level!(self.logger, level, "Error handling message{}; ignoring: {}", OptionalFromDebugger(&peer_node_id), e.err);
+											continue
+										},
+										msgs::ErrorAction::IgnoreDuplicateGossip => continue, // Don't even bother logging these
+										msgs::ErrorAction::IgnoreError => {
+											log_debug!(self.logger, "Error handling message{}; ignoring: {}", OptionalFromDebugger(&peer_node_id), e.err);
+											continue;
+										},
+										msgs::ErrorAction::SendErrorMessage { msg } => {
+											log_debug!(self.logger, "Error handling message{}; sending error message with: {}", OptionalFromDebugger(&peer_node_id), e.err);
+											self.enqueue_message($peer, &msg);
+											continue;
+										},
+										msgs::ErrorAction::SendWarningMessage { msg, log_level } => {
+											log_given_level!(self.logger, log_level, "Error handling message{}; sending warning message with: {}", OptionalFromDebugger(&peer_node_id), e.err);
+											self.enqueue_message($peer, &msg);
+											continue;
+										},
+									}
+								}
+							}
+						}
+					}
+
+					let mut peer_lock = peer_mutex.lock().unwrap();
+					let peer = &mut *peer_lock;
+					let mut msg_to_handle = None;
+					if peer_node_id.is_none() {
+						peer_node_id = peer.their_node_id.clone();
+					}
+
 					assert!(peer.pending_read_buffer.len() > 0);
 					assert!(peer.pending_read_buffer.len() > peer.pending_read_buffer_pos);
 
-					let mut read_pos = 0;
-					while read_pos < data.len() {
-						{
-							let data_to_copy = cmp::min(peer.pending_read_buffer.len() - peer.pending_read_buffer_pos, data.len() - read_pos);
-							peer.pending_read_buffer[peer.pending_read_buffer_pos..peer.pending_read_buffer_pos + data_to_copy].copy_from_slice(&data[read_pos..read_pos + data_to_copy]);
-							read_pos += data_to_copy;
-							peer.pending_read_buffer_pos += data_to_copy;
+					{
+						let data_to_copy = cmp::min(peer.pending_read_buffer.len() - peer.pending_read_buffer_pos, data.len() - read_pos);
+						peer.pending_read_buffer[peer.pending_read_buffer_pos..peer.pending_read_buffer_pos + data_to_copy].copy_from_slice(&data[read_pos..read_pos + data_to_copy]);
+						read_pos += data_to_copy;
+						peer.pending_read_buffer_pos += data_to_copy;
+					}
+
+					if peer.pending_read_buffer_pos == peer.pending_read_buffer.len() {
+						peer.pending_read_buffer_pos = 0;
+
+						macro_rules! insert_node_id {
```

Comment:
Do we need a macro for this? Propagating an Err response type seems fairly doable, including if this macro were to be converted to a regular function. Normally I wouldn't say anything, but since you're already touching this code :P

Reply:
This is quite disappointing. I guess this macro survives to live another day, but it's definitely in my crosshairs.